### PR TITLE
feat: devtoolsAtom

### PR DESCRIPTION
**Pull request by Aslemammad**
Feat #796 

This PR is also a WIP version of #575. Special thanks to @aulneau for his helpful code!

### CODE REVIEW

File: src/devtools/devtoolsAtom.ts [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+
+export function devtoolsAtom<Value>(
+  anAtom: PrimitiveAtom<Value>,
+  options?: Options
+): PrimitiveAtom<Value> {
+  const isWriteable = !!anAtom.write
+  const atomName = options?.name || anAtom.debugLabel || anAtom.toString()
+
+  const extension = getExtension()
+  const devtools = extension.connect({ name: atomName })
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:49]: We should do this in `onMount` or `write` of listenerAtom.

File: src/devtools/devtoolsAtom.ts [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+  let isTimeTraveling = false
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:51]: This is outside atom, which is shared among multiple instances. Please use `refAtom`. Refer `selectAtom`.

File: src/devtools/devtoolsAtom.ts [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    null,
+    (get, set, message: Message | { type: 'init' }) => {
+      if (message.type === 'init') {
+        return devtools.init(get(anAtom))
+      }
+
+      if (message.type === 'ACTION' && message.payload) {
+        try {
+          if (isWriteable) {
+            set(subscribeAtom, JSON.parse(message.payload))
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:63]: Is it intentional that this is not `anAtom`?

File: src/devtools/devtoolsAtom.ts [hunk 4]
```diff
@@ -0,0 +1,142 @@
...
+        }
+      } else if (message.type === 'DISPATCH' && message.state) {
+        if (
+          message.payload?.type === 'JUMP_TO_ACTION' ||
+          message.payload?.type === 'JUMP_TO_STATE'
+        ) {
+          isTimeTraveling = true
+
+          if (isWriteable) {
+            set(subscribeAtom, JSON.parse(message.state))
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:79]: Can this be `anAtom`? Then, we don't need `isTimeTraveling` at all?

File: src/devtools/devtoolsAtom.ts [hunk 5]
```diff
@@ -0,0 +1,142 @@
...
+        message.payload?.type === 'IMPORT_STATE'
+      ) {
+        const computedStates =
+          message.payload.nextLiftedState?.computedStates || []
+        computedStates.forEach(({ state }: { state: any }, index: number) => {
+          if (index === 0) {
+            devtools.init(state)
+          } else {
+            if (isWriteable) {
+              set(subscribeAtom, state)
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:98]: Same here. But, I may misunderstand something.

File: src/devtools/devtoolsAtom.ts [hunk 6]
```diff
@@ -0,0 +1,142 @@
...
+    setAtom({ type: 'init' })
+
+    return unsubscribe
+  }
+
+  const subscribeAtom: PrimitiveAtom<Value> = atom(
+    (get) => {
+      get(listenerAtom)
+
+      return anAtom.read(get)
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:120]: We should not use the internal `read` here. Can this be `get(anAtom)`?

File: src/devtools/devtoolsAtom.ts [hunk 7]
```diff
@@ -0,0 +1,142 @@
...
+    },
+    (get, set, update) => {
+      const writeReturn = anAtom.write(get, set, update)
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:123]: Please do `set(anAtom, update)`.

File: src/devtools/devtoolsAtom.ts [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+
+      if (isTimeTraveling) {
+        isTimeTraveling = false
+      } else {
+        devtools.send(
+          `${atomName} - ${new Date().toLocaleString()}`,
+          anAtom.read(get)
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:130]: Just send `writeReturn`.

File: src/devtools/devtoolsAtom.ts [hunk 9]
```diff
@@ -0,0 +1,142 @@
...
+        )
+      }
+
+      return writeReturn
+    }
+  )
+
+  subscribeAtom.onMount = anAtom.onMount
+  subscribeAtom.debugLabel = anAtom.debugLabel
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:139]: Don't copy these stuffs. onMount should just work. debugLabel should be set by a user.

File: src/devtools/devtoolsAtom.ts [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+
+  return subscribeAtom
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:141]: Not sure if there's a better name than `subscribeAtom`. 🤔

File: src/devtools/devtoolsAtom.ts [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+  } catch (e) {
+    throw new Error('Please install/enable Redux devtools extension')
+  }
+}
+
+export function devtoolsAtom<Value>(
+  anAtom: PrimitiveAtom<Value>,
+  options?: Options
+): PrimitiveAtom<Value> {
+  const isWriteable = !!anAtom.write
```
Comment (dai-shi) [inline src/devtools/devtoolsAtom.ts:45]: We don't need this flag, as we can assume `anAtom` is always writable.

**Review by dai-shi — COMMENTED** [review]
Overall, it looks good.
We can't write jest tests for this, can we?
Maybe mock devtools api?

File: src/devtools/devtoolsAtom.ts (hunk 10 above)
Comment (Aslemammad) [reply src/devtools/devtoolsAtom.ts:141]: newAtom?

File: src/devtools/devtoolsAtom.ts (hunk 8 above)
Comment (Aslemammad) [reply src/devtools/devtoolsAtom.ts:130]: sure! thinking about where write doesn't return anything.

Comment (Aslemammad) [reply src/devtools/devtoolsAtom.ts:130]: this doens't work, so `get(anAtom)` is better.

**Comment by Aslemammad** [comment]
@LuckyPigeon Could you test it to see if it works for atomFamily?

**Comment by LuckyPigeon** [comment]
@Aslemammad Sure. I can do that!

**Comment by dai-shi** [comment]
https://github.com/pmndrs/jotai/issues/796#issuecomment-956015136

Sorry @LuckyPigeon, we will start over.

**Pull request closed without merging**